File: scala-server/test/com/bolour/eagerwords/server/domain/ScorerSpec.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
 */

package com.bolour.eagerwords.server.domain

import com.bolour.eagerwords.server.util.WordUtil
import com.bolour.eagerwords.common.domain.{Piece, PlayPiece}
import com.bolour.plane.domain.Axis
import org.scalatest.{FlatSpec, Matchers}
import org.slf4j.LoggerFactory

class ScorerSpec extends FlatSpec with Matchers {
  val logger = LoggerFactory.getLogger(this.getClass)

  val dimension = 15
  val middle = dimension / 2
  val trayCapacity = 7

  def playInfo(word: String, strip: Strip): List[PlayPiece] = {
    ((0 until word.length) map { i: Int =>
      val ch = word(i)
      val point = strip.point(i)
      val moved = WordUtil.isBlankChar(strip.content(i))
      PlayPiece(Piece(ch, ""), point, moved)
    }).toList
  }

  val pointValues = List.fill(dimension, dimension)(1)

   val scorer = Scorer(dimension, trayCapacity, pointValues)

  "basic scoring with point values of 1" should "score correctly based on worth of letters" in {
    val info = playInfo("JOIN", Strip(Axis.X, 0, 0, 3, " O  "))
    logger.info(s"play pieces: ${info}")
    scorer.scoreWord(info) shouldEqual 3
  }

  // TODO. Add scoring tests with larger point values.
}

File: haskell-server/test/EagerWords/Server/Domain/ScorerSpec.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}

module EagerWords.Server.Domain.ScorerSpec where

import Test.Hspec

import EagerWords.Common.Domain.PlayPiece (PlayPiece, PlayPiece(PlayPiece), MoveInfo)
import EagerWords.Common.Domain.Piece (Piece, Piece(Piece))
import qualified EagerWords.Common.Domain.Piece as Piece
import qualified EagerWords.Server.Domain.Board as Board

import Bolour.Plane.Domain.Point (Point(Point))
import qualified Bolour.Plane.Domain.Axis as Axis
import qualified EagerWords.Server.Domain.Scorer as Scorer

pce :: Char -> Maybe Piece
pce s = Just $ Piece s "" -- Ignore id.

dimension = 5
trayCapacity = 3

baseGrid :: [[Maybe Piece]]
baseGrid = [
--        0        1        2        3        4
      [Nothing, Nothing, Nothing, Nothing, Nothing] -- 0
    , [pce 'C', pce 'A', pce 'R', Nothing, Nothing] -- 1
    , [Nothing, Nothing, Nothing, pce 'O', pce 'N'] -- 2
    , [pce 'E', pce 'A', pce 'R', Nothing, Nothing] -- 3
    , [Nothing, Nothing, Nothing, pce 'E', pce 'X'] -- 4
  ]

board = Board.mkBoardFromPieces baseGrid dimension

pointValues :: [[Int]]
pointValues = replicate dimension $ replicate dimension 1

scorer = Scorer.mkScorer pointValues
scorePlay = Scorer.scorePlay scorer

-- wt = Piece.letterWorth

-- toPlayPiece :: MoveInfo -> PlayPiece
-- toPlayPiece (ch, point, moved) = PlayPiece (Piece ch "") point moved
-- -- Note: piece id is not involved in scoring.

playPiece :: Char -> Int -> Int -> Bool -> PlayPiece
playPiece ch row col moved = PlayPiece (Piece ch "") (Point row col) moved

spec :: Spec
spec = do
  describe "calculate scores" $ do
    it "should find score for 1 move play" $ do
       let playPieces = [
                  playPiece 'N' 2 4 False
                , playPiece 'I' 3 4 True
                , playPiece 'X' 4 4 False
              ]
           score = scorePlay playPieces
       score `shouldBe` 1
    it "should find score for 2-move play" $ do
       let playPieces = [
                  playPiece 'S' 2 1 True
                , playPiece 'O' 2 2 True
                , playPiece 'O' 2 3 False
                , playPiece 'N' 2 4 False
              ]
           score = scorePlay playPieces
       score `shouldBe` 2





